lock/lock_windows.go

```diff
@@ -70,9 +70,5 @@ func winCreateEphemeral(name string) (windows.Handle, error) {
 		FILE_ATTRIBUTE_TEMPORARY  = 0x100
 		FILE_FLAG_DELETE_ON_CLOSE = 0x04000000
 	)
-	handle, err := windows.CreateFile(windows.StringToUTF16Ptr(name), 0, 0, nil, windows.OPEN_ALWAYS, FILE_ATTRIBUTE_TEMPORARY|FILE_FLAG_DELETE_ON_CLOSE, 0)
-	if err != nil {
-		return 0, err
-	}
-	return handle, nil
+	return windows.CreateFile(windows.StringToUTF16Ptr(name), 0, 0, nil, windows.OPEN_ALWAYS, FILE_ATTRIBUTE_TEMPORARY|FILE_FLAG_DELETE_ON_CLOSE, 0)
```

Comment:
On error, `CreateFileW` (the underlying call here) returns `INVALID_HANDLE_VALUE` (`windows.InvalidHandle`) (https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilew) which is defined to be `-1` not `0`: (https://referencesource.microsoft.com/#System.Runtime.Remoting/channels/ipc/win32namedpipes.cs,89)

I imagine this is why the `if err != nil { return 0, err }` is there, i.e. to normalize the return value to the zero value in the case of an error.

Though, this might not be a big deal, as the handle can be considered an undefined value in the event `err != nil` and so, should actually be rarely/never used. But this is a significant change of behavior that needs to be checked, and double-checked.

Reply:
>i.e. to normalize the return value to the zero value in the case of an error.

I wrote this originally and think that was the rationale.

Currently, the only caller `lockWindows`, immediately discards the `handle` if `err != nil`.
But I figure, if we're hoisting the error values, we may want to do the same with the system error.
And let the caller decide if any special action should be taken, or just bail out if they encounter any error.

A nice thing, is that this value is explicitly defined by the WINAPI to be an invalid handle, so even if it mistakenly propagates, passing that `handle` to `windows`/WINAPI methods should detect this, and return an error themselves saying the handle passed in was invalid.
*I think this is mostly true ^

That said, I don't think CreateFile will ever return a different value, we're more interested in `r2` the `GetLastError` value.
So it seems we can pass back both to the caller, or normalize `r1`.

I'm fine with either since `r2` seems most important here.

Reply:
:+1: sounds good. It’s also an unexported function, so we can actually fairly definitively make sure this won’t break any code.